tests/testthat/test-shape.R

```diff
@@ -83,3 +83,23 @@ test_that("shape_broadcast_() applies recycling rules", {
     class = "vctrs_error_incompatible_type"
   )
 })
+
+test_that("can combine shaped native classes (#1290, #1329)", {
+  x <- Sys.time() + c(1, 1e6)
+  dim(x) <- c(1, 2)
+  out <- vec_c(x, x)
+
+  expect_s3_class(out, c("POSIXct", "POSIXt"))
+  expect_dim(out, c(2, 2))
+
+  y <- Sys.time() + 1:3
+  dim(y) <- c(1, 3)
+
+  expect_snapshot(error = TRUE, vec_c(x, y))
+
+  d <- structure(Sys.Date(), dim = 1)
+  expect_equal(
+    vec_rbind(data.frame(d), data.frame(d)),
+    data.frame(d = structure(rep(Sys.Date(), 2), dim = 2))
+  )
+})
```

Comment:
I think this would be a useful test because it:

- tests this with factors (and factor level unioning)
- tests broadcasting is working
- tests dimension names are broadcast

``` r
x <- factor(c("x", "y", "z"))
dim(x) <- c(3, 1)
dimnames(x) <- list(c("r1", "r2", "r3"), "c1")

y <- factor(c("w", "x", "y", "z"))
dim(y) <- c(2, 2)

vec_cast(x, y)
#>    c1 c1
#> r1 x  x 
#> r2 y  y 
#> r3 z  z 
#> Levels: w x y z
```

Reply:
Nice one